modules/programs/i3blocks.nix

```
@@ -0,0 +1,109 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+
+let
+
+  cfg = config.programs.i3blocks;
+
+  # Re-make the atom type for the INI files.
+  # For some reason, the normal INI type seems to be incompatible with
+  # DAG
+  configAtomType = let
+    # Keep the INI atom type here
+    optType = with types; (nullOr (oneOf [ int bool str float ]));
+  in types.mkOptionType {
+    name = "INI config atom";
+    description = "INI atom (null, int, bool, string, or float)";
+    check = x: optType.check x;
+    merge = (loc: defs: (optType.merge loc defs));
+  };
+
+  # Create the type of the actual config type
+  configType = types.attrsOf configAtomType;
+
+  # The INI generator
+  mkIni = generators.toINI { };
+in {
+  meta.maintainers = [ hm.maintainers.noodlez1232 ];
+
+  options.programs.i3blocks = {
+    enable = mkEnableOption "i3blocks i3 status command scheduler";
+
+    package = mkOption {
+      type = types.package;
+      default = pkgs.i3blocks;
+      defaultText = literalExpression "pkgs.i3blocks";
+      description = "Package providing {command}`i3blocks`.";
+    };
+
+    bars = mkOption {
+      type = with types; attrsOf (hm.types.dagOf configType);
```

Sorry about the long delay!

I'm wondering if it wouldn't be sufficient to have
```suggestion
      type = with types; attrsOf (listOf configType);
```
here since I assume you typically full control over the block order.

---

I tried this for a bit, and lemme show you a snippet of how my HM config works, then it might make more sense why I chose a DAG.
```nix
let
# Host specific configuration
extraBottomBlocks = {
   main = {
      network = lib.hm.dag.entryBetween [ "date" ] [ "vol" ] {
         # Block here
      };
      # Other blocks here
   };
   framework = {
      battery = lib.hm.dag.entryBetween [ "date" ] [ "vol" ] {
         # Block here
      };
      # Other blocks here
   };
};
in
{
   programs.i3blocks = {
      bars = {
        top = {
           # ...
        };
        # Generic blocks
        bottom = {
            date = {
               # ...
            };
            vol = { };
        }; // extraBottomBlocks."${osConfig.networking.hostName}";
      };
   };
}
```

---

I had in mind the ability to set up kinda "generic" block structures and you can choose to put other blocks between them later on, making it a bit easier to work with in that way.

---

Also, I'd have to rewrite pretty much the entire module for this to work, which means a lot of redoing debugging and all that.

---

Fair enough, merged to master now 🙂